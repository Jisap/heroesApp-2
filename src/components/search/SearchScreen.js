import { useNavigate, useLocation } from "react-router-dom"
import queryString from "query-string"
import { useForm } from "../../hooks/useForms"
import { getHeroesByName } from "../../selectors/getHeroesByName"
import { HeroCard } from "../hero/HeroCard"
import { useMemo } from "react"


export const SearchScreen = () => {
    
    const navigate = useNavigate()
    const location = useLocation()  // hook que nos permite obtener un objeto con el pathname y los queryparams ( q=search )
    const { q = '' } = queryString.parse(location.search) // Separamos los queryParams  y nos quedamos solo con q que es el termino de la busqueda
   
    // Valores del form. - funcion que modifica esos valores
    const [ formValues, handleInputChange ] = useForm({searchText: q })
    const { searchText } = formValues   // Extraemos el valor del formulario que nos interesa, el searchText

    const handleSearch = (e) => {   // Cuando se haga click en el submit se llama a handleSearch y se construyen
        e.preventDefault()          // los query params en el pathname (http://localhost:3000/search?q=sdfdf)
        navigate(`?q=${searchText}`)
    }

    const heroesFiltered = useMemo(()=> getHeroesByName(q), [q])// construidos los query params y separado el termino de busqueda de ellos
                                                                // ejecutamos la busqueda  
    return (
        <>
           <h1>Búsquedas</h1>
           <hr />

           <div className="row">
               <div className="col-5">
                   <h4>Buscar</h4>
                   <hr />
                   <form onSubmit={ handleSearch }> 
                       <input
                            type="text"
                            placeholder="Buscar un héroe"
                            className="form-control"
                            name="searchText"
                            autoComplete="off"
                            value={ searchText }
                            onChange={ handleInputChange } // Cuando cambie este input llamamos a handleInputChange
                                                           // y cambiara el estado de formvalue -> searchText 
                        />

                        <div className="d-grid gap-2">
                            <button 
                                type="submit"
                                className="btn btn-outline-primary mt-3"
                                >
                                    Buscar...
                            </button>
                        </div>
                   </form>
               </div>

               <div className="col-7">
                   <h4>Resultados</h4>
                   <h4 />

                        {
                            (q === '')
                                ? <div className="alert alert-info">Buscar un héroe</div>
                                : (heroesFiltered.length === 0 ) && <div className="alert alert-danger">No hay resultados para: { q }</div>
                        }

                        {
                            heroesFiltered.map( hero => (
                                <HeroCard 
                                    key={ hero.id }
                                    { ...hero }
                                />
                            ) )
                        }

               </div>
           </div>
        </>
    )
}
