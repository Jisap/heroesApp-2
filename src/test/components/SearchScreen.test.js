import { mount } from "enzyme"
import { MemoryRouter } from "react-router-dom";
import { SearchScreen } from "../../components/search/SearchScreen"
// MemoryRouter nos permitirá mantener el mismo diseño que usamos con Routes

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'), // El mock usa el react-router-dom
    useNavigate: () => mockNavigate            // pero sobrescribe el useNavigate para hace un mock de el.      
}))


describe('Pruebas en <SearchScreen />', () => {
    
    test('debe de mostrarse correctamente con valores por defecto', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>
        )
        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.alert-info').text().trim()).toBe('Buscar un héroe')
    });

    test('debe de mostrar a Batman y el input con el valor del queryString ', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchScreen />
            </MemoryRouter>
        )

        expect( wrapper.find('input').prop('value')).toBe('batman')
    })
    
    test('debe de mostrar un error si no se encuentra el hero ', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=kdfjslkf']}>
                <SearchScreen />
            </MemoryRouter>
        )

        expect( wrapper.find('.alert-danger').text().trim()).toBe('No hay resultados para: kdfjslkf')
    })

    test('debe de llamar el navigate a la nueva pantalla ', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>
        )

        wrapper.find('input').simulate('change', {  // Simulamos que escribimos en el input batman
            target:{
                name: 'searchText',
                value: 'batman'
            }
        })

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        })

        expect( mockNavigate ).toHaveBeenCalledWith('?q=batman')
    })
    
})
