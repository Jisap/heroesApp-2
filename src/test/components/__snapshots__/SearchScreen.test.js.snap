// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> debe de mostrarse correctamente con valores por defecto 1`] = `
Array [
  <h1>
    Búsquedas
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un héroe"
          type="text"
          value=""
        />
        <div
          className="d-grid gap-2"
        >
          <button
            className="btn btn-outline-primary mt-3"
            type="submit"
          >
            Buscar...
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <h4 />
      <div
        className="alert alert-info"
      >
        Buscar un héroe
      </div>
    </div>
  </div>,
]
`;
